xscen/extract.py

```diff
@@ -862,119 +864,128 @@ def get_warming_level(
     """
-    if tas_baseline_period is None:
-        tas_baseline_period = ["1850", "1900"]
-    tas_baseline_period = standardize_periods(tas_baseline_period, multiple=False)
-
-    if tas_csv is None:
-        tas_csv = Path(__file__).parent / "data/IPCC_annual_global_tas.csv"
-
-    if isinstance(realization, xr.Dataset):
-        # get info on ds
-        if pd.isna(realization.attrs.get("cat:driving_model", None)):
-            source_ds = realization.attrs["cat:source"]
-        else:
-            institution_ds = realization.attrs.get("cat:driving_institution", None)
-            source_ds = realization.attrs["cat:driving_model"]
-            if (institution_ds is not None) and (
-                source_ds[0 : len(institution_ds)] == institution_ds
-            ):
-                source_ds = source_ds.replace(f"{institution_ds}-", "", 1)
-        exp_ds = realization.attrs["cat:experiment"]
-        member_ds = ".*" if ignore_member else realization.attrs["cat:member"]
-        mip_era_ds = realization.attrs["cat:mip_era"]
+    tas_baseline_period = standardize_periods(
+        tas_baseline_period or ["1850", "1900"], multiple=False
+    )
 
-        info_models = [f"{mip_era_ds}_{source_ds}_{exp_ds}_{member_ds}"]
+    if (window % 2) not in {0, 1}:
+        raise ValueError(f"window should be an integer, received {type(window)}")
 
-    elif isinstance(realization, str):
-        info_models = [realization]
-    elif isinstance(realization, list):
-        info_models = realization
-    else:
-        raise ValueError(
-            f"'realization' must be a Dataset, string or list. Received {type(realization)}."
-        )
+    FIELDS = ["mip_era", "source", "experiment", "member"]
 
-    # open csv
-    annual_tas = pd.read_csv(tas_csv, index_col="year")
-    out = dict()
-    for model in info_models:
-        if len(model.split("_")) != 4:
+    if tas_csv is None:
+        tas_csv = Path(__file__).parent / "data" / "IPCC_annual_global_tas.csv"
+
+    info = {}
+    if isinstance(realization, (xr.Dataset, str, dict)):
+        realization = [realization]
+    info_models = []
+    for real in realization:
+        info = {}
+        if isinstance(real, xr.Dataset):
+            attrs = get_cat_attrs(real)
+            # get info on ds
+            if attrs.get("driving_model") is None:
+                info["source"] = attrs["source"]
+            else:
+                info["source"] = attrs["driving_model"]
+            info["experiment"] = attrs["experiment"]
+            info["member"] = ".*" if ignore_member else attrs["member"]
+            info["mip_era"] = attrs["mip_era"]
+        elif isinstance(real, str):
+            (
+                info["mip_era"],
+                info["source"],
+                info["experiment"],
+                info["member"],
+            ) = real.split("_")
+        elif isinstance(real, dict) and set(real.keys()).issuperset(FIELDS):
+            info = real
```

Comment:
do we want to accept a dict with `driving_model` ?

Reply:
I'd say that the conversion from "driving_model" to "source" is to be done by the user when the dict is created.
This flexibility  for the dataset case is because it's easier than asking to convert the attribute name, I'm not sure I see the point for user-provided data.